src/main/scala/scalaz/stream/hash.scala

```diff
@@ -13,51 +15,84 @@ import process1._
  *   conditions. See $MutableProcess1 for more information.
  */
 object hash {
+  /** responsible for providing a MessageDigest for a type */
+  trait Digestable[A] {
+    def newDigester: MessageDigest
+  }
+
+  object Digestable {
+    def apply[A: Digestable] = implicitly[Digestable[A]]
+
+    def named[A](algorithm: String): Digestable[A] =
+      new Digestable[A] {
+        def newDigester =
+          MessageDigest.getInstance(algorithm)
+      }
+  }
+
+  sealed trait MD2
+  implicit val MD2Digestable = Digestable.named[MD2]("MD2")
+
+  sealed trait MD5
+  implicit val MD5Digestable = Digestable.named[MD5]("MD5")
+
+  sealed trait SHA1
+  implicit val SHA1Digestable = Digestable.named[SHA1]("SHA-1")
+
+  sealed trait SHA256
+  implicit val SHA256Digestable = Digestable.named[SHA256]("SHA-256")
+
+  sealed trait SHA384
+  implicit val SHA384Digestable = Digestable.named[SHA384]("SHA-384")
+
+  sealed trait SHA512
+  implicit val SHA512Digestable = Digestable.named[SHA512]("SHA-512")
+
   /**
    * Computes the MD2 hash of the input elements.
    * $MutableImpl
    */
-  val md2: Process1[ByteVector,ByteVector] = messageDigest("MD2")
+  val md2: Process1[ByteVector,ByteVector @@ MD2] = messageDigest[MD2]
 
   /**
    * Computes the MD5 hash of the input elements.
    * $MutableImpl
    */
-  val md5: Process1[ByteVector,ByteVector] = messageDigest("MD5")
+  val md5: Process1[ByteVector,ByteVector @@ MD5] = messageDigest[MD5]
 
   /**
    * Computes the SHA-1 hash of the input elements.
    * $MutableImpl
    */
-  val sha1: Process1[ByteVector,ByteVector] = messageDigest("SHA-1")
+  val sha1: Process1[ByteVector,ByteVector @@ SHA1] = messageDigest[SHA1]
 
   /**
    * Computes the SHA-256 hash of the input elements.
    * $MutableImpl
    */
-  val sha256: Process1[ByteVector,ByteVector] = messageDigest("SHA-256")
+  val sha256: Process1[ByteVector,ByteVector @@ SHA256] = messageDigest[SHA256]
 
   /**
    * Computes the SHA-384 hash of the input elements.
    * $MutableImpl
    */
-  val sha384: Process1[ByteVector,ByteVector] = messageDigest("SHA-384")
+  val sha384: Process1[ByteVector,ByteVector @@ SHA384] = messageDigest[SHA384]
 
   /**
    * Computes the SHA-512 hash of the input elements.
    * $MutableImpl
    */
-  val sha512: Process1[ByteVector,ByteVector] = messageDigest("SHA-512")
+  val sha512: Process1[ByteVector,ByteVector @@ SHA512] = messageDigest[SHA512]
 
-  private def messageDigest(algorithm: String): Process1[ByteVector,ByteVector] = {
-    def go(digest: MessageDigest): Process1[ByteVector,ByteVector] =
+  private def messageDigest[A: Digestable]: Process1[ByteVector,ByteVector @@ A] = {
```

Comment:
I made this private because `algorithm` could have been an arbitrary string and not only a valid algorithm name. Does it make sense to make this public now that you need to provide a `Digestable` instead of a `String`?